File: sdks/ts/src/clients/obligations/erc721/payment.ts
import { decodeAbiParameters, encodeAbiParameters, getAbiItem } from "viem";
import { abi as erc721PaymentAbi } from "../../../contracts/obligations/payment/ERC721PaymentObligation";
import { abi as atomicPaymentUtilsAbi } from "../../../contracts/utils/AtomicPaymentUtils";
import type { Erc721 } from "../../../types";
import { getAttestation, getAttestedEventFromTxHash, type ViemClient, writeContract } from "../../../utils";
import { getAtomicPaymentEscrowAttestation, type AtomicPaymentOptions } from "../../../utils/contractSafety";
import type { Erc721Addresses } from "./index";
import { makeErc721UtilClient } from "./util";

const erc721PaymentDecodeFunction = getAbiItem({
  abi: erc721PaymentAbi.abi,
  name: "decodeObligationData",
});

const erc721PaymentObligationDataType = erc721PaymentDecodeFunction.outputs[0];

/**
 * ERC721 Payment ObligationData type
 */
export type Erc721PaymentObligationData = {
  token: `0x${string}`;
  tokenId: bigint;
  payee: `0x${string}`;
};

/**
 * Encodes ERC721PaymentObligation.ObligationData to bytes.
 * @param data - struct ObligationData {address token, uint256 tokenId, address payee}
 * @returns abi encoded bytes
 */
export const encodeObligation = (data: Erc721PaymentObligationData): `0x${string}` => {
  return encodeAbiParameters([erc721PaymentObligationDataType], [data]);
};

/**
 * Decodes ERC721PaymentObligation.ObligationData from bytes.
 * @param obligationData - ObligationData as abi encoded bytes
 * @returns the decoded ObligationData object
 */
export const decodeObligation = (obligationData: `0x${string}`): Erc721PaymentObligationData => {
  return decodeAbiParameters([erc721PaymentObligationDataType], obligationData)[0] as Erc721PaymentObligationData;
};

export type Erc721PaymentClient = ReturnType<typeof makeErc721PaymentClient>;

export const makeErc721PaymentClient = (viemClient: ViemClient, addresses: Erc721Addresses) => {
  const util = makeErc721UtilClient(viemClient, addresses);

  const getSchema = async () =>
    await viemClient.readContract({
      address: addresses.paymentObligation,
      abi: erc721PaymentAbi.abi,
      functionName: "ATTESTATION_SCHEMA",
      authorizationList: undefined,
    });

  return {
    address: addresses.paymentObligation,
    atomicPaymentUtilsAddress: addresses.atomicPaymentUtils,
    getSchema,

    encodeObligationRaw: (data: { token: `0x${string}`; tokenId: bigint; payee: `0x${string}` }) => {
      return encodeAbiParameters([erc721PaymentObligationDataType], [data]);
    },

    encodeObligation: (token: Erc721, payee: `0x${string}`) => {
      return encodeAbiParameters(
        [erc721PaymentObligationDataType],
        [
          {
            token: token.address,
            tokenId: token.id,
            payee,
          },
        ],
      );
    },

    decodeObligation: (obligationData: `0x${string}`) => {
      return decodeAbiParameters([erc721PaymentObligationDataType], obligationData)[0];
    },

    getObligation: async (uid: `0x${string}`) => {
      const [attestation, schema] = await Promise.all([getAttestation(viemClient, uid, addresses), getSchema()]);

      if (attestation.schema !== schema) {
        throw new Error(`Unsupported schema: ${attestation.schema}`);
      }
      const data = decodeAbiParameters([erc721PaymentObligationDataType], attestation.data)[0];

      return {
        ...attestation,
        data,
      };
    },

    pay: async (
      price: Erc721,
      payee: `0x${string}`,
      refUID: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000",
    ) => {
      const hash = await writeContract(viemClient, {
        address: addresses.paymentObligation,
        abi: erc721PaymentAbi.abi,
        functionName: "doObligation",
        args: [
          {
            token: price.address,
            tokenId: price.id,
            payee,
          },
          refUID,
        ],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },

    approveAndPay: async (
      price: Erc721,
      payee: `0x${string}`,
      refUID: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000",
    ) => {
      await util.approve(price, "payment");
      const hash = await writeContract(viemClient, {
        address: addresses.paymentObligation,
        abi: erc721PaymentAbi.abi,
        functionName: "doObligation",
        args: [
          {
            token: price.address,
            tokenId: price.id,
            payee,
          },
          refUID,
        ],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },

    /**
     * Security note: uses AtomicPaymentUtils, which was not included in the
     * professional manual audits and has only been reviewed by automated audit
     * tooling so far.
     */
    payErc721AndCollect: async (escrowUid: `0x${string}`, options?: AtomicPaymentOptions) => {
      await getAtomicPaymentEscrowAttestation(viemClient, addresses, escrowUid, options);
      const hash = await writeContract(viemClient, {
        address: addresses.atomicPaymentUtils,
        abi: atomicPaymentUtilsAbi.abi,
        functionName: "payErc721AndCollect",
        args: [escrowUid],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },
  };
};

File: sdks/rs/src/clients/obligations/erc721/payment.rs
//! ERC721 payment obligation client

use alloy::primitives::{Address, FixedBytes};
use alloy::rpc::types::TransactionReceipt;
use alloy::sol_types::SolValue;

use crate::contracts;
use crate::types::{ApprovalPurpose, DecodedAttestation, Erc721Data};
use crate::utils::contract_safety::ensure_packaged_escrow_attester;

use super::Erc721Module;

/// Payment API for ERC721 tokens
pub struct Payment<'a> {
    module: &'a Erc721Module,
}

impl<'a> Payment<'a> {
    pub fn new(module: &'a Erc721Module) -> Self {
        Self { module }
    }

    /// Get the contract address
    pub fn address(&self) -> Address {
        self.module.addresses.payment_obligation
    }

    /// Gets a payment obligation by its attestation UID.
    pub async fn get_obligation(
        &self,
        uid: FixedBytes<32>,
    ) -> eyre::Result<
        DecodedAttestation<contracts::obligations::ERC721PaymentObligation::ObligationData>,
    > {
        let eas_contract =
            contracts::IEAS::new(self.module.addresses.eas, &self.module.wallet_provider);

        let attestation = eas_contract.getAttestation(uid).call().await?;
        let obligation_data =
            contracts::obligations::ERC721PaymentObligation::ObligationData::abi_decode(
                &attestation.data,
            )?;

        Ok(DecodedAttestation {
            attestation,
            data: obligation_data,
        })
    }

    /// Makes a direct payment with ERC721 tokens.
    pub async fn pay(
        &self,
        price: &Erc721Data,
        payee: Address,
    ) -> eyre::Result<TransactionReceipt> {
        let payment_obligation_contract = contracts::obligations::ERC721PaymentObligation::new(
            self.module.addresses.payment_obligation,
            &self.module.wallet_provider,
        );

        let receipt = payment_obligation_contract
            .doObligation(
                contracts::obligations::ERC721PaymentObligation::ObligationData {
                    token: price.address,
                    tokenId: price.id,
                    payee,
                },
                FixedBytes::<32>::ZERO,
            )
            .send()
            .await?
            .get_receipt()
            .await?;

        Ok(receipt)
    }

    /// Makes a direct payment with ERC721 tokens after approving the token transfer.
    pub async fn approve_and_pay(
        &self,
        price: &Erc721Data,
        payee: Address,
    ) -> eyre::Result<(TransactionReceipt, TransactionReceipt)> {
        let util = self.module.util();
        let approval_receipt = util.approve(price, ApprovalPurpose::Payment).await?;
        let payment_receipt = self.pay(price, payee).await?;
        Ok((approval_receipt, payment_receipt))
    }

    /// Pays an ERC721 payment obligation and collects the matching escrow atomically.
    ///
    /// Security note: uses AtomicPaymentUtils, which has not been included in
    /// professional manual audits and has only been reviewed by automated audit
    /// tooling so far.
    pub async fn pay_erc721_and_collect(
        &self,
        escrow_uid: FixedBytes<32>,
    ) -> eyre::Result<TransactionReceipt> {
        self.ensure_supported_atomic_payment_escrow(escrow_uid)
            .await?;
        self.pay_erc721_and_collect_unchecked(escrow_uid).await
    }

    /// Pays an ERC721 payment obligation and collects without SDK escrow-attester validation.
    ///
    /// Use only after independently validating that `escrow_uid` was authored by
    /// the escrow contract you intend to settle.
    pub async fn pay_erc721_and_collect_unchecked(
        &self,
        escrow_uid: FixedBytes<32>,
    ) -> eyre::Result<TransactionReceipt> {
        let utility = contracts::utils::AtomicPaymentUtils::new(
            self.module.addresses.atomic_payment_utils,
            &self.module.wallet_provider,
        );

        Ok(utility
            .payErc721AndCollect(escrow_uid)
            .send()
            .await?
            .get_receipt()
            .await?)
    }

    async fn ensure_supported_atomic_payment_escrow(
        &self,
        escrow_uid: FixedBytes<32>,
    ) -> eyre::Result<()> {
        let eas = contracts::IEAS::new(self.module.addresses.eas, &self.module.wallet_provider);
        let escrow = eas.getAttestation(escrow_uid).call().await?;
        ensure_packaged_escrow_attester(escrow.attester, &self.module.packaged_escrow_obligations)
    }
}
